src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs

```diff
@@ -0,0 +1,377 @@
+            TypeDesc maybeTypeDesc;
+            foreach (MethodDesc md in targetType.GetMethods())
+            {
+                // Check the target and current method match static/instance state.
+                if (context.IsTargetStatic != md.Signature.IsStatic)
+                {
+                    continue;
+                }
+
+                // Check for matching name
+                if (!md.Name.Equals(name))
+                {
+                    continue;
+                }
+
+                // Validate calling convention.
+                if ((MethodSignatureFlags.UnmanagedCallingConventionMask & md.Signature.Flags)
+                    != (MethodSignatureFlags.UnmanagedCallingConventionMask & context.Declaration.Signature.Flags))
+                {
+                    continue;
+                }
+
+                // Validate the return type and prepare for validating
+                // the current signature's argument list.
+                int sigCount = context.Declaration.Signature.Length;
+                if (context.Kind == UnsafeAccessorKind.Constructor)
+                {
+                    if (!md.Signature.ReturnType.IsVoid)
+                    {
+                        continue;
+                    }
+                }
+                else
+                {
+                    declTypeDesc = context.Declaration.Signature.ReturnType;
+                    maybeTypeDesc = md.Signature.ReturnType;
+                    if (declTypeDesc != maybeTypeDesc)
+                    {
+                        continue;
+                    }
+
+                    // Non-constructor accessors skip the first argument
+                    // when validating the target argument list.
+                    sigCount--;
+                }
+
+                // Validate argument count matches.
+                if (sigCount != md.Signature.Length)
+                {
+                    continue;
+                }
+
+                // Validate arguments match - reverse order
+                for (; sigCount > 0; --sigCount)
+                {
+                    declTypeDesc = context.Declaration.Signature[sigCount];
+                    maybeTypeDesc = md.Signature[sigCount];
+                    if (declTypeDesc != maybeTypeDesc)
+                    {
+                        break;
+                    }
+                }
+
+                // If we validated all arguments, we have a match.
+                if (sigCount != 0)
+                {
+                    continue;
+                }
+
+                context.TargetMethod = md;
+                return true;
+            }
+            return false;
+        }
+
+        private static bool TrySetTargetMethodCtor(ref GenerationContext context)
+        {
+            // Special case the default constructor case.
+            if (context.Declaration.Signature.Length == 0
+                && context.TargetType.HasExplicitOrImplicitDefaultConstructor())
```

Comment:
HasExplicitOrImplicitDefaultConstructor will return true for valuetypes without a default constructor and we end up setting TargetMethod to null below. Why do we need this special case?

Reply:
> Why do we need this special case?

We don't. I thought it was going to be useful, but perhaps it isn't.